go/vt/vttablet/tabletserver/repltracker/writer_test.go

```diff
@@ -47,16 +48,222 @@ func TestWriteHeartbeat(t *testing.T) {
+
+	tw := newTestWriter(db, nil, tabletenv.Disable, 0)
+
+	// Even though disabled, the writer will have an on-demand duration set.
+	assert.Equal(t, defaultOnDemandDuration, tw.onDemandDuration)
+
+	db.AddQueryPattern("^INSERT INTO.*", &sqltypes.Result{})
+
+	writes.Reset()
+	writeErrors.Reset()
+
+	tw.writeHeartbeat()
+	lastWrites := writes.Get()
+	assert.EqualValues(t, 1, lastWrites)
+	assert.EqualValues(t, 0, writeErrors.Get())
+
+	t.Run("closed, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+	tw.Open()
+	defer tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.NotNil(t, rateLimiter)
+	}
+	t.Run("open, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	t.Run("request heartbeats, heartbeats", func(t *testing.T) {
+		tw.RequestHeartbeats()
+		ctx, cancel := context.WithTimeout(context.Background(), tw.onDemandDuration-time.Second)
+		defer cancel()
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				assert.EqualValues(t, 0, writeErrors.Get())
+				currentWrites := writes.Get()
+				assert.Greater(t, currentWrites, lastWrites)
+				lastWrites = currentWrites
+			}
+		}
+	})
+	t.Run("exhaust lease, no heartbeats", func(t *testing.T) {
+		<-time.After(tw.onDemandDuration)
+		currentWrites := writes.Get()
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, currentWrites, writes.Get())
+	})
+	tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+}
+
+// TestWriteHeartbeatOnDemand tests that the heartbeat writer initiates leased heartbeats once opened,
+// and then upon RequestHeartbeats().
+func TestWriteHeartbeatOnDemand(t *testing.T) {
+	defaultOnDemandDuration = 7 * time.Second
+	onDemandDuration := 3 * time.Second
+
+	db := fakesqldb.New(t)
+	defer db.Close()
+
+	tw := newTestWriter(db, nil, tabletenv.Heartbeat, onDemandDuration)
+
+	assert.Equal(t, onDemandDuration, tw.onDemandDuration)
+
+	db.AddQueryPattern("^INSERT INTO.*", &sqltypes.Result{})
+
+	writes.Reset()
+	writeErrors.Reset()
+
+	tw.writeHeartbeat()
+	lastWrites := writes.Get()
+	assert.EqualValues(t, 1, lastWrites)
+	assert.EqualValues(t, 0, writeErrors.Get())
+
+	t.Run("closed, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+	tw.Open()
+	defer tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.NotNil(t, rateLimiter)
+	}
+	t.Run("open, initial heartbeats", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), tw.onDemandDuration-time.Second)
```

Comment:
Any reason to use the `-time.Second` part? That would put us at 2 seconds and it seems like it may be possible that we then only with the ticker once before the context is done.

Reply:
Yes. The reasoning is that I don't want to get past the expiration of the heartbeat lease. This function is checking:

- That heartbeats are generated when on-demand is engaged (upon `Open()` in this case)
- And that they _keep_ being generated for the duration of the on-demand lease. The heartbeat interval in this test is `250ms`, and we test every 1 second, just shy of the expiration.

If we go past the expiration, then we can wrongly condlude that heartbeats are not being produced when they should (they shouldn't, past the expiration).